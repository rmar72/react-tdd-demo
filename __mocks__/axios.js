const axiosMock = jest.genMockFromModule('axios');

let mockResponse = {
    data: {
        "shops":
            [{
                "location": "tst address",
                "address": "test address"
            }]
    }
};

axiosMock.get.mockImplementation(req);

function req(){
    return new Promise(function(resolve){
        axiosMock.delayTimer = setTimeout( function(){
            resolve(mockResponse);
        }, 1000);
    });
}

module.exports = axiosMock;