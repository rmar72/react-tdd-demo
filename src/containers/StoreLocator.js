import React, { Component } from 'react';
import Header from '../components/Header';
import Button from '../components/Button';
import Map from '../components/Map';
import mapChooser from '../mapChooser';
import axios from 'axios';

class StoreLocator extends Component {
    constructor(props){
        super(props);
        this.state = {
            currentMap: 'default.png',
            shops: []
        }
        
        this.chooseMap = this.chooseMap.bind(this);
    }

    async componentDidMount(){
        let response = await axios.get('http://localhost:3000/data/shops.json');
        this.setState({shops: response.data.shops}, ()=> console.log(this.state.shops))
    }

    chooseMap(e){
        this.setState({currentMap: mapChooser(e.target.value)});
    }

    render(){
        let storeButtons = this.state.shops.map((shop, id)=>{
            return (
                <Button
                    key={id}
                    location={shop.location}
                    handleClick={this.chooseMap}
                />
            );
        })
        return(
            <div>
                <Header />
                <div className="buttons">
                    {storeButtons}
                </div>
                <Map imagename={this.state.currentMap} />
            </div>
        );
    }
}

export default StoreLocator;