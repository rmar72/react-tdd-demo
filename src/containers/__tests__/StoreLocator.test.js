import React from 'react';
import { shallow } from 'enzyme';
import StoreLocator from '../StoreLocator';
import axios from 'axios';
import renderer from 'react-test-renderer';

describe('StoreLocator', () => {
    let mountedStoreLocator;
    // create mountedStoreLocator, thats a shallow render of the Component, so I can use it in each of my tests
    beforeEach( () => {
        mountedStoreLocator = shallow(<StoreLocator />);
    });

    it('renders correctly', () => {
        const tree = renderer.create(<StoreLocator />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('it calls axios.get in componentDidMount', () => {
        // we do return here bc we want to do a network call on didMount, so we have to wait for it to return,
        // hmmm still a bit ambigous about it having to return
        return mountedStoreLocator.instance().componentDidMount().then(()=>{
            // then we do the test
            expect(axios.get).toHaveBeenCalled();
        });
    });

    it('calls axios.get with correct url', () => {
        return mountedStoreLocator.instance().componentDidMount().then(()=>{
            expect(axios.get).toHaveBeenCalledWith('http://localhost:3000/data/shops.json');
        });
    });

    it('update state with api data', () => {
        return mountedStoreLocator.instance().componentDidMount().then(()=>{
            expect(mountedStoreLocator.state()).toHaveProperty('shops', 
                [
                    {
                        "location": "tst address",
                        "address": "test address"
                    }
                ]
            );
        });
    });

    it('renders without crashing', () => {
        mountedStoreLocator = shallow(<StoreLocator />);
    });

    it('renders Header', () => {
        const headers = mountedStoreLocator.find('Header');
        expect(headers.length).toBe(1);
    });

    // no longer working since we changed the flow, there is async call for data that buttons rely on
    // it('renders three buttons', () => {
    //     const buttons = mountedStoreLocator.find('Button');
    //     expect(buttons.length).toBe(3);
    // });

    it('renders a map', () => {
        const map = mountedStoreLocator.find('Map');
        expect(map.length).toBe(1);
    });
});

describe('Choose Map', () => {
    it('updates this.state.currentMap using the location passed to it', () => {
        let mountedStoreLocator = shallow(<StoreLocator />);
        let mockEventData = {target: {value:'testland'}};
        mountedStoreLocator.instance().chooseMap(mockEventData); // chooseMap is an instance fn, that sets the state, its not clear tho
        // I see it now, so after red bar, we have to ccreate a fn chooseMap that is going to setState, and if we pass the mockEventData
        // it will update currentMap, it's just confusing as to when you need to mock fns and when you dont
        expect(mountedStoreLocator.instance().state.currentMap).toBe('testland.png');

    });
});