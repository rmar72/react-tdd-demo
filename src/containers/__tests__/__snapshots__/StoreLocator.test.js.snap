// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreLocator renders correctly 1`] = `
<div>
  <div>
    <img
      alt="wired-coffee brand logo"
      src="images/logo.png"
    />
  </div>
  <div
    className="buttons"
  />
  <div
    className="MapBox"
  >
    <img
      src="images/default.png"
    />
  </div>
</div>
`;
